BRIDGE-636 Redesign tapping activity to support bilateral tapping task

syoung-smallwisdom opened the pull request:
This is a fairly significant change to AppCore. I added unit testing around adding in an activity to test tapping with each hand as well as archiving that data. I removed APCDataArchiver (it is not referenced anywhere) and added APCTaskResultArchiver. This base class is the default archiver for APCBaseTaskViewController.  

Please let me know if you need a tour. It's a lot of changes.

syoung-smallwisdom committed:
Added convenience method for flipping an animated UIImage.

syoung-smallwisdom committed:
BRIDGE-636 Added bilateral tapping activity

syoung-smallwisdom committed:
BRIDGE-636 Add unit tests for bilateral tapping task step creation

syoung-smallwisdom committed:
BRIDGE-636 Refactor to support fixing unit tests and to prepare archiving for being testable

syoung-smallwisdom committed:
BRIDGE-636 Refactor task result archiving to use a helper class APCTaskResultArchiver

EricSiegNW commented:
The bilateral tapping activity code all looked fine. I haven't wrapped my head around the implications/necessity of the APCTaskResultArchiver yet and how it compares to APCDataArchiverAndUploader yet.

Erin-Mounts commented on APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m:
```diff
@@ -33,41 +33,24 @@
     }
     
-    return [fileResultIdentifier stringByAppendingString:stepIdentifier];
+    return [result copy];
```
why not just return result?

syoung-smallwisdom replied:
result is mutable and the method signature is immutable.

Erin-Mounts replied:
oh yeah derp.

> On Dec 14, 2015, at 4:46 PM, Shannon Young notifications@github.com wrote:
> 
> In APCAppCore/APCAppCore/Library/Categories/ORKFileResult+Filename.m https://github.com/Sage-Bionetworks/AppCore/pull/53#discussion_r47582729:
> 
> > ```
> >  }
> > ```
> > -    return [fileResultIdentifier stringByAppendingString:stepIdentifier];
> > -    return [result copy];
> >   result is mutable and the method signature is immutable.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/Sage-Bionetworks/AppCore/pull/53/files#r47582729.

Erin-Mounts commented on APCAppCore/APCAppCore/Library/Categories/ORKOrderedTask+APCHelper.m:
```diff
@@ -0,0 +1,157 @@
+        [orkTask.steps[tappingInstructionIdx] isKindOfClass:[ORKInstructionStep class]])
+    {
+        tappingInstructionStep = (ORKInstructionStep *)orkTask.steps[tappingInstructionIdx];
+        tappingInstructionStep.detailText = NSLocalizedStringWithDefaultValue(@"Step Detail Text", @"APCAppCore", APCBundle(),
```
I'd use a more specific key to avoid potential collisions in the future.

syoung-smallwisdom commented:
@EricSiegNW 
WRT 'APCDataArchiverAndUploader' verse 'APCDataArchiverAndUploader', they were doing two different things. It is _very_ confusing. 'APCDataArchiver' was unused so I deleted it. It duplicated code in 'APCBaseTaskViewController' called by '-(void)archiveResults' to take a 'ORKTaskResult' and put the information from that into a freshly instantiated 'APCDataArchive' (not to be confused with 'APCDataArchiver'). Then the base class implementation for archiving a tapping result or a memory game result, were empty methods. There was an assumption that subclasses of 'APCBaseViewController' used to implement a memory game or tapping activity would archive the results themselves. This didn't really make sense so I moved the logic for doing that into the base implementation. I moved to using the 'APCTaskResultArchiver' for simpler unit test injection and to simplify any subclassing hierarchy implemented by the apps that use this.

Erin-Mounts commented on APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.h:
```diff
@@ -89,7 +89,18 @@
  @param     filename                Filename for the json data to be included without path extension
  */
-- (void)insertIntoArchive:(NSDictionary *)dictionary filename: (NSString *)filename;
+- (void)insertDictionaryIntoArchive:(NSDictionary *)dictionary filename: (NSString *)filename;
```
better name is better

syoung-smallwisdom replied:
Yeah, I got very confused about why it wasn't crashing on attempting to push a dictionary when expecting NSData.

Erin-Mounts commented:
one nitpicky comment on a localization key, one dumb question you already answered, otherwise looks great.

syoung-smallwisdom committed:
Updated copyright info

syoung-smallwisdom committed:
Changed localization key
